### PR TITLE
[SPARK-17764][SQL] Add `to_json` supporting to convert nested struct column to JSON string

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to add `to_json` function in contrast with `from_json` in Scala, Java and Python.

It'd be useful if we can convert a same column from/to json. Also, some datasources do not support nested types. If we are forced to save a dataframe into those data sources, we might be able to work around by this function.

The usage is as below:

``` scala
val df = Seq(Tuple1(Tuple1(1))).toDF("a")
df.select(to_json($"a").as("json")).show()
```

``` bash
+--------+
|    json|
+--------+
|{"_1":1}|
+--------+
```
## How was this patch tested?

Unit tests in `JsonFunctionsSuite` and `JsonExpressionsSuite`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @marmbrus Could you take a look please?

**Comment by HyukjinKwon** [comment]
interesting. It's unicode only for Python 3.4? I will look into this deeper.

**Comment by HyukjinKwon** [comment]
Ah, this was a known issue

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -1729,6 +1729,29 @@ def from_json(col, schema, options={}):
...
+    >>> from pyspark.sql import Row
+    >>> from pyspark.sql.types import *
+    >>> data = [(1, Row(name='Alice', age=2))]
+    >>> df = spark.createDataFrame(data, ("key", "value"))
+    >>> df.select(to_json(df.value).alias("json")).collect()
+    [Row(json=u'{"age":2,"name":"Alice"}')]
+    """
+
+    sc = SparkContext._active_spark_context
+    jc = sc._jvm.functions.to_json(_to_java_column(col), options)
```
Comment (holdenk) [inline python/pyspark/sql/functions.py]: This is super minor, but there is a pretty consistent pattern for all of the other functions here (including `from_json`), it might be good to follow that same pattern for consistencies sake since there isn't an obvious reason why that wouldn't work here.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1729,6 +1729,29 @@ def from_json(col, schema, options={}):
...
 
 
+@ignore_unicode_prefix
+@since(2.1)
+def to_json(col, options={}):
+    """
+    Converts a column containing a [[StructType]] into a JSON string. Returns `null`,
+    in the case of an unsupported type.
+
+    :param col: struct column
```
Comment (holdenk) [inline python/pyspark/sql/functions.py]: Would    `:param col: name of column containing the struct` maybe be more consistent with the other pydocs for the functions? (I only skimmed a few though so if its the other way around thats cool).

**Review by holdenk — COMMENTED** [review]
Thanks for working on this :) Just did a quick look at the Python parts and some minor style changes that we might want to consider :)

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: Sure, let me try to double check other comments as well.

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: @holdenk Thank you for your comment. Could you please a bit elaborate this comment? I am a bit not sure on what to fix.

Comment (holdenk) [inline python/pyspark/sql/functions.py]: actually nvm my original comment, the more I look at this file the less it seems the pattern is overly consistent and this same pattern is done elsewhere within the file.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala [hunk 3]
```diff
@@ -343,4 +343,23 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+    val schema = StructType(StructField("a", IntegerType) :: Nil)
+    val struct = Literal.create(create_row(1), schema)
+    checkEvaluation(
+      StructToJson(Map.empty, struct),
+      """{"a":1}"""
+    )
+  }
+
+  test("to_json - invalid type") {
+    val schema = StructType(StructField("a", CalendarIntervalType) :: Nil)
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: Hmm, I realize this is a little different than `from_json`, but it seems it would be better to eagerly throw an `AnalysisException` to say the schema contains an unsupported type.  We know that ahead of time, and otherwise its kind of mysterious why all the values come out as `null`.

**Review by marmbrus — CHANGES_REQUESTED** [review]
Overall, looks pretty good.  What do you think about failing earlier and more obviously when there are unsupported datatypes?

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala (hunk 3 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala]: I would like to leave a note. In case of CSV we are verifying the types before actually running tasks but for JSON it is not doing this. So, I make this `SparkSQLJsonProcessingException` which is technically a `RuntimeException`. However, if you want me to fix it here (adding a logic to verify the schema ahead) I will definitely do this here together.

**Comment by felixcheung** [comment]
hi - where are we on this?

**Comment by HyukjinKwon** [comment]
@felixcheung Thank you for pinging!

@marmbrus Do you please mind if I handle checking the schema ahead for csv/json in read/write in another PR?

**Comment by marmbrus** [comment]
It would be really nice to fail in analysis rather than execution.  What if it only fails after hours of computation? As a user I'd be upset.  I'm also concerned they will think it's a spark bug.

**Comment by HyukjinKwon** [comment]
@marmbrus Sure (I didn't mean I am not going to do this..), I just handled the case in this PR for `to_json`.

BTW, I would like to note that there are the same problems in other JSON related functionalities.  For example, I might have to add 

```scala
override def checkInputDataTypes(): TypeCheckResult = {
  ...
 JacksonUtils.verifySchema(child.dataType.asInstanceOf[StructType])
  ...
}
```

for `from_json` as well. Let me please open a follow up for adding this logic and test for JSON related functionalities.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala [hunk 4]
```diff
@@ -29,4 +31,28 @@ object JacksonUtils {
...
+
+      case at: ArrayType => verifyType(at.elementType)
+
+      case mt: MapType => verifyType(mt.keyType)
+
+      case udt: UserDefinedType[_] => verifyType(udt.sqlType)
+
+      case _ =>
+        throw new UnsupportedOperationException(
+          s"JSON conversion does not support to process ${dataType.simpleString} type.")
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonUtils.scala]: `does not support to process` is a little hard to parse.  Maybe `Unable to convert column ${name} of type ${dataType.simpleString} to JSON.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala [hunk 5]
```diff
@@ -123,8 +122,9 @@ private[sql] class JacksonGenerator(
     case _ =>
       (row: SpecializedGetters, ordinal: Int) =>
         val v = row.get(ordinal, dataType)
-        sys.error(s"Failed to convert value $v (class of ${v.getClass}}) " +
-          s"with the type of $dataType to JSON.")
+        throw new SparkSQLJsonProcessingException(
+          s"Failed to convert value $v (class of ${v.getClass}}) " +
+            s"with the type of $dataType to JSON.")
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonGenerator.scala]: I would avoid this change since its throwing a private exception type to the user now.

**Review by marmbrus — CHANGES_REQUESTED** [review]
Only minor comments.  Thanks for working on this!

**Comment by HyukjinKwon** [comment]
@marmbrus Thank you so much.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Let me take a look into this deeper if some same tests constantly fail.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
It seems the test is not related with this PR.

**Comment by felixcheung** [comment]
looks good, should we clarify output JSON is in JSON Lines format? http://jsonlines.org/

**Comment by HyukjinKwon** [comment]
I understand it might be good to mention (and I have thought for a while) but I feel a bit hesitant because we could express json strings in dataframe are stored without newline separators in a way (I am not nitpicking).

The function itself produces each json string which can be the standard json and held in memory. This might not be applied to the standard (it says "text file format"..).

I am actually rather neutral. We could also mention that too just to simply say it is in-line. I am going to submit another PR to deal with the schema verification for other JSON functionalities and if we should mention this, we should do it for other JSON functionalities (such as `Dataset.toJSON`). Maybe we could talk about this further if it sounds okay.

**Comment by HyukjinKwon** [comment]
Oh nvm. It seems it'd be better to mention.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 6]
```diff
@@ -2936,6 +2936,51 @@ object functions {
...
+   *
+   * @group collection_funcs
+   * @since 2.1.0
+   */
+  def to_json(e: Column, options: java.util.Map[String, String]): Column =
+    to_json(e, options.asScala.toMap)
+
+  /**
+   * Converts a column containing a [[StructType]] into a JSON string
+   * ([[http://jsonlines.org/ JSON Lines text format or newline-delimited JSON]]) with the
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: I don't think that this case really follows "JSON lines".  It is a string inside of a larger dataframe.  There are no newlines involved.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 7]
```diff
@@ -494,3 +495,46 @@ case class JsonToStruct(schema: StructType, options: Map[String, String], child:
...
+  @transient
+  lazy val gen =
+    new JacksonGenerator(child.dataType.asInstanceOf[StructType], writer)
+
+  override def dataType: DataType = StringType
+  override def children: Seq[Expression] = child :: Nil
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (StructType.acceptsType(child.dataType)) {
+      try {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Sorry, one final comment as I'm looking at this more closely.  I don't think we should use exceptions for control flow in the common case.  Specifically, `verifySchema` should work the same way as `acceptsType` above and return a boolean.

**Review by marmbrus — CHANGES_REQUESTED** [review]
I think we are getting close.  Just one more large comment and some doc changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 7 above)
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Ah, yes, makes sense but if `verifySchema` returns a boolean, we could not find which field and type are problematic.

Maybe, I can make do one of the below:
  - this logic in `verifySchema` into `checkInputDataTypes`
  - `verifySchema` returns the unsupported fields. and types.
  - Just fix the exception message without the information of unsupported fields and types.

If you pick one, I will follow (or please let me know if there is a better way)!

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by marmbrus** [comment]
Thanks, I'm going to merge this to master.

**Comment by HyukjinKwon** [comment]
Thank you for merging this!
